server/src/main/java/com/thoughtworks/go/server/service/RuleAwarePluginProfileService.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2019 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.server.service;
+
+
+import com.thoughtworks.go.config.ConfigTag;
+import com.thoughtworks.go.config.RuleAwarePluginProfile;
+import com.thoughtworks.go.config.RuleAwarePluginProfiles;
+import com.thoughtworks.go.config.exceptions.GoConfigInvalidException;
+import com.thoughtworks.go.config.exceptions.RecordNotFoundException;
+import com.thoughtworks.go.config.update.RuleAwarePluginProfileCommand;
+import com.thoughtworks.go.domain.config.ConfigurationProperty;
+import com.thoughtworks.go.plugin.api.response.validation.ValidationResult;
+import com.thoughtworks.go.server.domain.Username;
+import com.thoughtworks.go.server.service.result.LocalizedOperationResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.thoughtworks.go.i18n.LocalizedMessage.entityConfigValidationFailed;
+import static com.thoughtworks.go.i18n.LocalizedMessage.saveFailedWithReason;
+
+public abstract class RuleAwarePluginProfileService<M extends RuleAwarePluginProfile> {
+    protected final GoConfigService goConfigService;
+    protected final EntityHashingService hashingService;
+    protected Logger LOGGER = LoggerFactory.getLogger(getClass());
+
+    public RuleAwarePluginProfileService(GoConfigService goConfigService, EntityHashingService hashingService) {
+        this.goConfigService = goConfigService;
+        this.hashingService = hashingService;
+    }
+
+    protected abstract RuleAwarePluginProfiles<M> getPluginProfiles();
+
+    public M findProfile(String id) {
+        return getPluginProfiles().find(id);
+    }
+
+    private void validatePluginProperties(RuleAwarePluginProfileCommand command, RuleAwarePluginProfile newPluginProfile) {
+        try {
+            ValidationResult result = command.validateUsingExtension(newPluginProfile.getPluginId(), newPluginProfile.getConfiguration().getConfigurationAsMap(true));
+            addErrorsToConfiguration(result, newPluginProfile);
+        }catch (RecordNotFoundException e) {
+            newPluginProfile.addError("pluginId", String.format("Plugin with id `%s` is not found.", newPluginProfile.getPluginId()));
+        }catch (Exception e) {
```

Comment:
Might have to add https://github.com/gocd/gocd/blob/master/server/src/main/java/com/thoughtworks/go/server/service/PluginProfilesService.java#L69

Reply:
Done.